### ISSUE TITLE
Crash: unowned self in fetched results delegate completion

### ISSUE THREAD

**Issue by LeeMatthewHiggins** [jessesquires/JSQDataSourcesKit#17]
I get a crash:

```
* thread #1: tid = 0x30b3fb, 0x000000010046ec70 libswiftCore.dylib`swift_retainUnowned + 44, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x10046ec70)
  * frame #0: 0x000000010046ec70 libswiftCore.dylib`swift_retainUnowned + 44
    frame #1: 0x00000001002bb65c JSQDataSourcesKit`JSQDataSourcesKit.CollectionViewFetchedResultsDelegateProvider.(finished=Swift.Bool at 0x000000016fda6370, self=0x0000000174058630).getter : JSQDataSourcesKit.(BridgedFetchedResultsDelegate in _301B319B6B7C8F32E76BCF594821AAF4)).(closure #4).(closure #2) + 40 at FetchedResultsDelegate.swift:106
    frame #2: 0x00000001002bb814 JSQDataSourcesKit`reabstraction thunk helper <T_0_0> from @callee_owned (@unowned Swift.Bool) -> (@unowned ()) to @callee_owned (@in Swift.Bool) -> (@out ()) + 44 at FetchedResultsDelegate.swift:0
    frame #3: 0x00000001002bb84c JSQDataSourcesKit`reabstraction thunk helper <T_0_0> from @callee_owned (@in Swift.Bool) -> (@out ()) to @callee_owned (@unowned Swift.Bool) -> (@unowned ()) + 44 at FetchedResultsDelegate.swift:0
    frame #4: 0x00000001002bb898 JSQDataSourcesKit`reabstraction thunk helper from @callee_owned (@unowned Swift.Bool) -> (@unowned ()) to @callee_unowned @objc_block (@unowned ObjectiveC.ObjCBool) -> (@unowned ()) + 64 at FetchedResultsDelegate.swift:0
    frame #5: 0x0000000187ece110 UIKit`-[UICollectionView _updateAnimationDidStop:finished:context:] + 1472
    frame #6: 0x000000018837272c UIKit`__34-[UICollectionViewAnimation start]_block_invoke223 + 264
    frame #7: 0x0000000187d2855c UIKit`-[UIViewAnimationBlockDelegate _didEndBlockAnimation:finished:context:] + 408
    frame #8: 0x0000000187d280c4 UIKit`-[UIViewAnimationState sendDelegateAnimationDidStop:finished:] + 188
    frame #9: 0x0000000187d27fcc UIKit`-[UIViewAnimationState animationDidStop:finished:] + 104
    frame #10: 0x000000018763162c QuartzCore`CA::Layer::run_animation_callbacks(void*) + 296
```

In FetchedResultsDelegate.swift
line 98

I made the code safe here by making the completion's 'self' weak. This fixes the problem......

```
 didChangeContent: { [unowned self] (controller) -> Void in

            self.collectionView?.performBatchUpdates({ () -> Void in
                self.applyObjectChanges()
                self.applySectionChanges()
            },
            completion:{ [weak self] (finished) -> Void in

                if let strongSelf = self
                {
                    if strongSelf.sectionChanges.count > 0 {
                        // if sections have changed, reload to update supplementary views
                        strongSelf.collectionView?.reloadData()
                    }

                    strongSelf.sectionChanges.removeAll()
                    strongSelf.objectChanges.removeAll()
                }
                else
                {
                    println("Self sometimes nil as view controller is animated away on pop")
                }
            })
```

After investigation is seems that view controllers are dealloc'd BEFORE their related views, it seems the collection view is then calling its delegate after the controller has gone away (along with the datasource provider).

**Comment by jessesquires** [comment]
Thanks @LeeMatthewHiggins ! :smile: 

I suspect you should set your collectionView `delegate` and `dataSource` to `nil` your view controller's `deinit`. Can you confirm if that fixes this instead?

Related discussion here: https://github.com/jessesquires/JSQMessagesViewController/issues/201

**Comment by LeeMatthewHiggins** [comment]
This was the first thing I tired, it did not fix the issue. :(

I think the BridgedFetchedResultsDelegate  is living a bit longer that it should or maybe theres a dispatch thats running on the next runloop after things have been released.

**Comment by jessesquires** [comment]
Ok cool. What are the exact steps to reproduce the issue?

**Comment by LeeMatthewHiggins** [comment]
So basically I have Async background updates to coredata that can then merge at any point. So the main UI context can get updates whenever (all thread safe, used in many apps)
I have a menu with items (on a FRC), you tap on an item and it pushes a new controller with the item (and kicks off an async refresh of the model layer), the items each have subitems on a fetched results controller (FRC). 
All I have to do is jump in and out of the items. about 40-45% of the time I will get the crash. ATM I have a very dumb cache, so coredata is getting updated a lot more than it will in production.

It seems that if the items in the FRC are updating when the controller is animating off the screen I get the crash. (this happens a lot with my dumb cache so thats why i’m getting it a lot)
so it is an edge case normally.

I noticed the example code is not async so it might not have been exercised in this way.

It is however working very well with the temp “fix” I did.
